src/Lykke.AlgoStore.Services/Validation/CSharpCodeBuildSession.cs

```diff
@@ -0,0 +1,309 @@
+                throw new ArgumentNullException(nameof(code));
+
+            _code = code;
+            _sourceText = SourceText.From(code);
+        }
+
+        public async Task<ValidationResult> Validate()
+        {
+            if (_syntaxValidationResult != null)
+                return _syntaxValidationResult;
+
+            _syntaxTree = CSharpSyntaxTree.ParseText(_code);
+
+            var validationMessages = new List<ValidationMessage>(_syntaxTree.GetDiagnostics()
+                                                                            .Select(DiagnosticToValidationMessage));
+
+            if (ErrorExists(validationMessages))
+                return CreateAndSetValidationResult(out _syntaxValidationResult, false, validationMessages);
+
+            var root = (CompilationUnitSyntax)await _syntaxTree.GetRootAsync();
+
+            var syntaxWalker = new CSharpAlgoValidationWalker(_sourceText);
+            syntaxWalker.Visit(root);
+
+            validationMessages.AddRange(syntaxWalker.GetMessages());
+
+            if(ErrorExists(validationMessages))
+                return CreateAndSetValidationResult(out _syntaxValidationResult, false, validationMessages);
+
+            // Semantic validation
+
+            var coreLib = MetadataReference.CreateFromFile(typeof(object).Assembly.Location);
+            var fxLibs = _allowedFxLibs
+                .Select(l => MetadataReference.CreateFromFile(
+                    Assembly.Load(
+                        $"{l.Item1}, Version={l.Item2}, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a").Location))
+                .ToArray();
+
+
+            _compilation = CSharpCompilation.Create("CodeValidation")
+                            .AddSyntaxTrees(_syntaxTree)
+                            .AddReferences(coreLib)
+                            .AddReferences(fxLibs)
+                            .AddReferences(await NuGetReferenceProvider.GetReferences());
+
+            _compilation = _compilation.WithOptions(new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+
+            _semanticModel = _compilation.GetSemanticModel(_syntaxTree, false);
+            var semanticDiagnostics = _semanticModel.GetDiagnostics();
+
+            validationMessages.AddRange(semanticDiagnostics.Select(DiagnosticToValidationMessage));
+
+            return CreateAndSetValidationResult(out _syntaxValidationResult, 
+                                                !ErrorExists(validationMessages), 
+                                                validationMessages);
+        }
+
+        public async Task<AlgoMetaDataInformation> ExtractMetadata()
+        {
+            var root = (CompilationUnitSyntax)await _syntaxTree.GetRootAsync();
+            var namespaceDeclaration = (NamespaceDeclarationSyntax)root.Members[0];
+            var classDeclaration = namespaceDeclaration.Members
+                .OfType<ClassDeclarationSyntax>()
+                .First();
```

Comment:
There's no way to know if the algo class will be in a namespace or if it will be the first class in that namespace.
It would be better to expose the algo class node from the syntax walker and use that instead